### PR TITLE
Show download progress/indicator while exporting to avoid multiple clicks

### PR DESCRIPTION
**Pull request by wilsonpg**
###### Before submitting the PR, please make sure you do the following

- [x] If you're attempting to fix a translation issue, please submit your changes to our [POEditor project](https://poeditor.com/join/project/ynjQmwSsGh) instead of opening a PR. (N/A)

### Tests

- [x] Run the frontend and Cypress end-to-end tests with `yarn lint && yarn test`)
- [x] If there are changes to the backend codebase, run the backend tests with `clojure -X:dev:test` (N/A)
- [x] Sign the [Contributor License Agreement](https://docs.google.com/a/metabase.com/forms/d/1oV38o7b9ONFSwuzwmERRMi9SYrhYeOrkbmNaq9pOJ_E/viewform)
      (unless it's a tiny documentation change).

#### Summary

This PR is in reference to #10566 

This feature was important to me because I was concerned about using metabase in production without it.  I’m from a tech-savvy demographic and I found myself clicking the download button multiple times, basically assuming it was broken or something because there was no loading indicator.  This user behavior sent multiple requests and could reasonably threaten the stability of the server.  This PR simply closes the download popover, disables the download button / uses the already existing loading spinner as an icon for the widget, and then downloads the file asynchronously so that it can report progress back to the download widget and stop the spinner / reenable the button.  It should properly inform the user that the download has commenced while simultaneously preventing them from triggering another download during that time.

#### Decision Analysis

There's a few points here that might not be immediately clear why I approached them the way that I did:

1. In order to report on the progress of the download, I needed to abstract the logic to a custom handleSubmit function

This means we're straying a bit from the default html form behavior and also that we have to replicate some logic to push the download through:

- Send a request with the encoded form data as the body (POST) or the query (GET)
- Use the response to create a mock anchor tag and click it (thus triggering the download behavior)

2. The Triggerable Popover didn’t appear to have a way to manually close it

I originally was hiding this while the download was pending (using the isOpen prop), but it felt like a strange user experience for the popover to open again once the download was finished.  Not necessarily the worst option, but I expected it to stay closed.

By default, the popover only closes when you click outside or press escape and using the isOpen prop works like a controlled value.  For now, I’ve mocked an ESCAPE keypress to auto-close the download popover when the handleSubmit event begins.

3. What on earth is that code that’s creating the filename

It’s just there to ensure that the filename standard that y’all have employed remains the same.  For whatever reason, the stream/blob process removes this name.  This PR retrieves it from the response header and parsing it into the previous query_result[DATE TIME].extension format, so this isn’t really a change.

#### UI Screenshots

I tested this on both POST requests (downloading from a question - view footer) and GET requests (downloading from an embedded dashboard using JWT), and it functions well in both scenarios.  Please see below to see an example of the spinner in the UI:

Pasted Graphic

Pasted Graphic 1

I can’t think of anything else that requires discussion, but please let me know if you have any additional questions!

### CODE REVIEW

File: frontend/src/metabase/query_builder/components/QueryDownloadWidget.jsx [hunk 1]
```diff
@@ -162,17 +186,37 @@ const DashboardEmbedQueryButton = ({
...
+  if ([`idle`, `resolved`, `rejected`].includes(status)) {
+    return (
+      <Tooltip tooltip={t`Download full results`}>
+        <Icon title={t`Download this data`} name={icon} size={20} />
+      </Tooltip>
+    );
+  } else if (status === "pending") {
+    return (
+      <Tooltip tooltip={t`Downloading...`}>
+        <LoadingSpinner />
```
Comment (mazameli) [inline frontend/src/metabase/query_builder/components/QueryDownloadWidget.jsx:212]: The spinner just looks a bit big right now. This little tweak makes it look more like it fits in with the other icons (screenshot below).

```suggestion
        <LoadingSpinner size={18} />
```

image

File: frontend/src/metabase/query_builder/components/QueryDownloadWidget.jsx [hunk 2]
```diff
@@ -162,17 +186,37 @@ const DashboardEmbedQueryButton = ({
...
+const renderIcon = ({ icon, status }) => {
+  if ([`idle`, `resolved`, `rejected`].includes(status)) {
+    return (
+      <Tooltip tooltip={t`Download full results`}>
+        <Icon title={t`Download this data`} name={icon} size={20} />
+      </Tooltip>
+    );
+  } else if (status === "pending") {
+    return (
+      <Tooltip tooltip={t`Downloading...`}>
```
Comment (mazameli) [inline frontend/src/metabase/query_builder/components/QueryDownloadWidget.jsx:211]:
```suggestion
      <Tooltip tooltip={t`Downloading…`}>
```

**Review by mazameli — CHANGES_REQUESTED** [review]
Thanks for the PR! Just two small corrections please.

**Review by mazameli — APPROVED** [review]

**Review by daltojohnso — CHANGES_REQUESTED** [review]
Hi there! Thanks for submitting this PR -- it is a much improved UX over what currently exists. I've added a few change requests to your code. Let me know if you need further clarification on what I've written.

**Review by daltojohnso — APPROVED** [review]
LGTM! This will merge into master from https://github.com/metabase/metabase/pull/21037

**Pull request closed without merging**